Mp 90 profile model

BsBou opened the pull request:
Written basic test for profile model
Test by running `bundle exec rspec spec/models`
Please give feedback!

adrianHards commented on app/models/user.rb:
```diff
@@ -4,7 +4,7 @@ class User < ApplicationRecord
   validates :email, uniqueness: true
   validates :role, presence: true
   validate :password_regex
```
Not seen this before, how does this work?

BsBou replied:
What are you asking about here?
The only added line is `has_many :profiles`

adrianHards replied:
sorry, think comment goes beneath quoted line by default but still confusing 'validate :password_regex', might not even be your code though!

adrianHards commented on spec/models/profile_spec.rb:
```diff
@@ -1,4 +1,35 @@
+
+  describe "Validations" do
+    it "Has a valid linkedin url" do
+      expect { create(:profile, linkedin: 'Boris Johnson') }.to raise_error(ActiveRecord::RecordInvalid)
```
nice! Only improvement would be to run against the expected error, e.g. raise_error("Validation failed: store can't be negative"); not sure what the quick way of doing this would be though, is there a way we can list all possible errors?

adrianHards replied:
here's a quick example I've just used in mine:

      context "title too long" do
        it "is not valid with too long a title" do
          subject.title = ("a" * 51).to_s
          expect { subject.save! }.to raise_error(ActiveRecord::RecordInvalid, "Validation failed: Title is too long (maximum is 50 characters)")
        end
      end

adrianHards commented on spec/models/profile_spec.rb:
```diff
@@ -1,4 +1,35 @@
+    
+    it "has a profile picture" do
+      profile = (build(:profile))
+      file = URI.open("https://images.unsplash.com/photo-1613679074971-91fc27180061?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1887&q=80")
```
using URI.open makes it a file of sorts?

adrianHards commented on spec/models/profile_spec.rb:
```diff
@@ -1,4 +1,35 @@
+    it "has a profile picture" do
+      profile = (build(:profile))
+      file = URI.open("https://images.unsplash.com/photo-1613679074971-91fc27180061?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1887&q=80")
+      profile.photo.attach(io: file, filename: "avatar.jpg")
```
nice solution to image testing, assuming it reflects what Cloudinary does as well. Found this to understand what you were doing, could refactor slightly: https://guides.rubyonrails.org/v7.0.0/active_storage_overview.html

CraigNorford replied:
As a profile picture will be added after the creation of the profile do we want to enforce it in the unit test? or do you want to rename the test to "can have photo attached"?

adrianHards commented on spec/models/profile_spec.rb:
```diff
@@ -1,4 +1,35 @@
+
+    it "must belong to a user" do
+      profile = (build(:profile))
+      expect(profile.user).to be_valid
```
see use of shoulda gem in my open PR, might be worth using as well?

adrianHards reviewed (commented):
we'll all need to be a bit more consistent with how we write tests, but looks good and I learnt a lot from reading it

BsBou commented:
- Added validations to pass existing tests
- Profiles are valid if bio, job, linked, github are nil. Added `allow_nil: true` to model validations and wrote tests for it.
- Refactored tests

adrianHards commented on app/models/profile.rb:
```diff
@@ -1,3 +1,8 @@
 class Profile < ApplicationRecord
+  validates :job, format: {with: /\A[a-zA-Z0-9 ]+\z/}, allow_nil: true
+  validates :linkedin, format: { with: /.*linkedin.\S*\/\S*/ }, allow_nil: true
+  validates :github, format: { with: /.*github.\S*\/\S*/ }, allow_nil: true
```
fancy! I assume we allow nil initially and then they fill in? I assume profile is created on user creation? I guess this gets at what I was asking earlier (re value of profile) but not relevant to this pr

adrianHards commented on spec/models/profile_spec.rb:
```diff
@@ -5,11 +5,9 @@
-    end
-    
-    it "has a profile picture" do
+      expect { create(:profile, user: nil) }.to raise_error(ActiveRecord::RecordInvalid, "Validation failed: User must exist")
```
looks good

adrianHards commented on config/database.yml:
```diff
@@ -6,7 +6,7 @@
 default: &default
   adapter: postgresql
-  host: localhost
+  # host: localhost
```
why does it need to be commented out? Just curious

adrianHards replied:
why does it need to be commented out? Just curious

adrianHards reviewed (changes requested):
I would recommend creating an initial subject where everything is valid. Small addition. See: 

    it "is valid with valid attributes" do
      expect(subject).to be_valid
    end
    
    in https://github.com/Mentor-Pathway/mentor-platform/pull/49/files#diff-c028eaf1a5d3be209fb2bbc438de4f76e29d034bc312931b9a04ada252a82a43

adrianHards commented on spec/models/profile_spec.rb:
```diff
@@ -19,17 +17,42 @@
+        expect { create(:profile, linkedin: 'Boris Johnson') }.to raise_error(ActiveRecord::RecordInvalid, "Validation failed: Linkedin is invalid")
+      end
+      it "github url is a valid url" do
+        expect { create(:profile, github: 'torvalds') }.to raise_error(ActiveRecord::RecordInvalid, "Validation failed: Github is invalid")
```
```
  describe "create a challenge" do
    subject {
      described_class.new(
        user: create(:user),
        title: "Roman numeral to integer",
        details: "Convert Roman numerals to numbers and convert numbers to Roman numerals"
      )
    }

    it "is valid with valid attributes" do
      expect(subject).to be_valid
    end
```

adrianHards commented on spec/models/profile_spec.rb:
```diff
@@ -17,6 +17,10 @@
   describe "Validations" do
+    it 'is valid with valid attributes' do
+      profile = (build(:profile))
+      expect(profile).to be_valid
```
does this work? Should there always be a user for this to be valid, or am I misunderstanding? Or another way of putting it, does build profile create an empty profile with no user?

adrianHards replied:
nevermind, didn't know how factory worked
